streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1721,40 +1724,33 @@ object ZStream extends ZStreamPlatformSpecific {
   ): ZStream[R, E, A] =
     new ZStream[R, E, A] {
-      override def fold[R1 <: R, E1 >: E, A1 >: A, S]: Fold[R1, E1, A1, S] =
-        ZManaged.succeed { (s, cont, g) =>
-          for {
-            output  <- Queue.bounded[Take[E1, A1]](outputBuffer).toManaged(_.shutdown)
-            runtime <- ZIO.runtime[R].toManaged_
-            eitherStream <- UIO {
-                             register(
-                               k =>
-                                 runtime.unsafeRunAsync_(
-                                   k.foldCauseM(
-                                     _.failureOrCause match {
-                                       case Left(None)    => output.offer(Take.End).unit
-                                       case Left(Some(e)) => output.offer(Take.Fail(Cause.fail(e))).unit
-                                       case Right(cause)  => output.offer(Take.Fail(cause)).unit
-                                     },
-                                     a => output.offer(Take.Value(a)).unit
-                                   )
+      def fold[R1 <: R, E1 >: E, A1 >: A, S]: Fold[R1, E1, A1, S] = foldDefault
+
+      override def process: ZManaged[R, E, InputStream[E, A]] =
+        for {
+          output  <- Queue.bounded[InputStream[E, A]](outputBuffer).toManaged(_.shutdown)
+          runtime <- ZIO.runtime[R].toManaged_
+          eitherStream <- UIO(
+                           register(
+                             k =>
+                               runtime.unsafeRunAsync_(
+                                 k.foldCauseM(
+                                   _.failureOrCause match {
+                                     case Left(None)    => output.offer(IO.fail(None)).unit
+                                     case Left(Some(e)) => output.offer(IO.fail(Some(e))).unit
+                                     case Right(cause)  => output.offer(IO.halt(cause)).unit
+                                   },
+                                   a => output.offer(UIO.succeed(a)).unit
                                  )
-                             )
-                           }.toManaged_
-            s <- eitherStream match {
-                  case Right(stream) =>
-                    output.shutdown.toManaged_ *>
-                      stream.fold[R1, E1, A1, S].flatMap(fold => fold(s, cont, g))
-                  case Left(canceler) =>
-                    ZStream
-                      .fromQueue(output)
-                      .unTake
-                      .fold[R1, E1, A1, S]
-                      .flatMap(fold => fold(s, cont, g))
-                      .ensuring(canceler)
-                }
-          } yield s
-        }
+                               )
+                           )
+                         ).toManaged_
+          is <- eitherStream match {
+                 case Left(canceler) =>
+                   ZManaged.succeed(output.take.flatten).ensuring(UIO.effectTotal(println("canceler")) *> canceler)
```

Comment:
Rogue `println` here

Reply:
Yes, this code seemed dubious to me when I wrote it, but it actually works properly. I'll remove it of course.

Reply:
Read it as: "when we exit the scope defined by this InputStream, be sure to run this canceller"

Reply:
Yesss, we are running only the inner program in a loop, the scope stays the same. Great!

Reply:
Done.